### ISSUE TITLE
[BUG]Output very loud at the beginning  when  aplay via headset(0,0) or media playback(0,4) from Runtime PM status  suspend mode

### ISSUE THREAD

**Issue by Liviali155** [thesofproject/linux#1463]
**Describe the bug**
Output very loud at the beginning when aplay via headset(0,0) or media playback(0,4) from Runtime PM status suspend mode 

**To Reproduce**
1.sudo reboot
2.cat "/sys/bus/pci/devices/0000:`lspci | grep -i audio | cut -d ' ' -f 1`/power/runtime_status"
suspend
3.aplay -Dhw:0,4 -r48000 -c2 -fs16_le wav/sine_48k_stereo_440.wav  -vv -i  or
   aplay -Dhw:0,0 -r48000 -c2 -fs16_le wav/sine_48k_stereo_440.wav  -vv -i

**Reproduction Rate**
Tried 10times can reproduce 10times

**Expected behavior**
The sound at the beginning should not become so  loud

**Impact**
The sound at the beginning become very loud suddenly and strident

Tested on 20190929，issue can also be reproduced，details as below
Firmware: 8528d932 https://github.com/thesofproject/sof/commits/master
Kernel: f1c5cf5e https://github.com/thesofproject/linux/commits/topic/sof-dev

<img width="960" alt="Capture" src="https://user-images.githubusercontent.com/56707932/68575776-cb97e380-04a7-11ea-9da5-800734e79ce3.PNG">
<img width="960" alt="Capture1" src="https://user-images.githubusercontent.com/56707932/68575779-cc307a00-04a7-11ea-85d0-522ca9dfaac5.PNG">

**Environment**
Platform: APL UP2 with I2S PCM512x
Firmware: fb068e6a https://github.com/thesofproject/sof/commits/master
Kernel: 4425a3e5 https://github.com/thesofproject/linux/commits/topic/sof-dev
Topology: file: tools/topology/sof-apl-pcm512x.tplg same as firmware

**sof error trace**
```
CORE  LEVEL      COMP_ID                TIMESTAMP            DELTA                FILE_NAME	CONTENT
    0      1           SA          49214787.135417  49214788.000000      src/lib/agent.c:55  	validate(), ll drift detected, delta = 22690
```

[sine_48k_stereo_440.zip](https://github.com/thesofproject/linux/files/3830574/sine_48k_stereo_440.zip)

[dmesgnoise.log](https://github.com/thesofproject/linux/files/3830747/dmesgnoise.log)
[sof-errortrace.log](https://github.com/thesofproject/linux/files/3834029/sof-errortrace.log)
[softrace.log](https://github.com/thesofproject/linux/files/3834016/softrace.log)

**Comment by sinahuang** [comment]
Issue still can be reproduced on TGL RVP with onboard ACL711 in SDW mod and ALC1308 in I2S mode.

scenario1: aplay -Dhw:0,0 -r48000 -c2 -fs16_le cut.wav -vv -i-->the audio become louder at the beginning.

scenario2:aplay -Dhw:0,0 -r48000 -c2 -fs16_le cut.wav -vv -i-->s3-->wake up-->the audio become louder at the beginning.

**Test recipe**

Kernel:https://github.com/thesofproject/linux/commits/integration/soundwire-latest   commit: 2933f73
FW: https://github.com/thesofproject/sof/commits/master commit: d1dcdfe
Tplg: Same with TGL FW branch commit: sof-tgl-rt711-rt1308-nohdmi.tplg rename to sof-tgl-rt711-rt1308.tplg
Platform:TGL RVP with onboard ACL711 in SDW mod and ALC1308 in I2S mode.

[demsg.log](https://github.com/thesofproject/linux/files/3834937/demsg.log)
[sof_error_trace.log](https://github.com/thesofproject/linux/files/3834938/sof_error_trace.log)
[sof_trace.log](https://github.com/thesofproject/linux/files/3834940/sof_trace.log)

**Comment by dengyangchao** [comment]
Issue can not reproduce on:
CFL-S RVP with onborad codec ALC700 in HDA mode
ICL RVP with onborad codec ALC700 in HDA mode
GLK Chrome with onbord codec DA7219 in I2S mode
APL UP2  nocodec 
BYT MB  nocodec
BYT MB  rt5682 in I2S mode

Issue can reproduce on:
APL UP2 with codec PCM512X in I2S mode
TGL RVP with onboard ALC711 in SDW mod and ALC1308 in I2S mode
CML SDW Laptop with ALC711+ ALC1308+ ALC715

**Comment by plbossart** [comment]
@dengyangchao  I don't understand your results and the description of the bug.

On APL Up2, there is a single output, so we cannot change between media and headset. The bug description is misleading, I have no idea what to look for.

On SDW platforms, you did you clarify on which output the sound is 'too loud'? Is this a problem with the speaker outputs? with the headset? both?
And btw the headset and speakers are completely independent, so when you switch from one to the other, it's equivalent to waiting for the same output to go through a pm_runtime suspend/resume cycles (typically need 5s of idle).

Please clarify what the problem really is and how you observe it, thanks.

**Comment by dengyangchao** [comment]
@plbossart 
1.For APL UP2 change aplay between media(hw0,4) and headset(hw0,0)  actual is hw 0,4 <=>hw0,0.
2.For TGL RVP with onboard ALC711 in SDW mod and ALC1308 in I2S mode, hw0,0 is headset, and TGL can reproduce the issue in 2 scenarios, when pm_runtime from idle to active or wake up from S3 during aplay headset(hw0,0).

**Comment by Liviali155** [comment]
I retested on APL UP2 with codec PCM512X in I2S mode,the problem is 
Output very loud at the beginning  when  aplay via (0,0) or media playback(0,4) from Runtime PM status  suspend mode

**Comment by plbossart** [comment]
Sorry, I don't get the issue. Your wording combines changes to the output as well as suspend/resume, that's two separate cases which can be tested separately

I think you should first focus on suspend-resume and describe the issue there.

The change from outputs does not happen usually at the alsa hardware device level.

**Comment by Liviali155** [comment]
> Sorry, I don't get the issue. Your wording combines changes to the output as well as suspend/resume, that's two separate cases which can be tested separately
> 
> I think you should first focus on suspend-resume and describe the issue there.
> 
> The change from outputs does not happen usually at the alsa hardware device level.

1.sudo reboot
2. wait the runtime_status is suspend，then aplay via （0,0) or (0,4)media playback can reproduce this issue
if the runtime_status is still active ，then you aplay via (0,0) or (0,4) media play back，can not reproduce this issue

**Comment by plbossart** [comment]
@Liviali155 does this happen when you play from /dev/zero ?

@shumingfan can you comment here?
Wondering if the io_init does something that could cause the pop noise.

**Comment by Liviali155** [comment]
> @Liviali155 does this happen when you play from /dev/zero ?
I tested using /dev/zero，it actually has  pop noise 
<img width="960" alt="Capture" src="https://user-images.githubusercontent.com/56707932/68912724-fcbc3080-0793-11ea-9059-829d9df6907e.PNG">

**Comment by shumingfan** [comment]
> @Liviali155 does this happen when you play from /dev/zero ?
> 
> @shumingfan can you comment here?
> Wondering if the io_init does something that could cause the pop noise.

I checked the dmesg log from @sinahuang .
It doesn't redo the io_init function. I also try to test on my side with Dell machine, but I don't hear the pop noise while playing from /dev/zero.
@Liviali155  Could you do the loopback test before sending data to codec using sine wav file?
like this, Playback data -> DSP output -> return to DSP input -> Capture data
We could check the capture data if it is correct or not.

**Comment by Liviali155** [comment]
> > @Liviali155 does this happen when you play from /dev/zero ?
> > @shumingfan can you comment here?
> > Wondering if the io_init does something that could cause the pop noise.
> 
> I checked the dmesg log from @sinahuang .
> It doesn't redo the io_init function. I also try to test on my side with Dell machine, but I don't hear the pop noise while playing from /dev/zero.
> @Liviali155 Could you do the loopback test before sending data to codec using sine wav file?
> like this, Playback data -> DSP output -> return to DSP input -> Capture data
> We could check the capture data if it is correct or not.

@shumingfan I tested  with /dev/zero can  not hear the pop noise,but when I do the loopback test,it actually has pop noise, the screenshot is from capture data when I do loopback test

**Comment by shumingfan** [comment]
@Liviali155 The method of loopback test is the same with what I say? Is it loopback in intel side?
If yes, the loud sound should come from DSP output.

**Comment by Liviali155** [comment]
> @Liviali155 The method of loopback test is the same with what I say? Is it loopback in intel side?
> If yes, the loud sound should come from DSP output.

Previous loopback is after sending data to code. Error not seen on nocodec. 

Here is the result on nocodec.
aplay  sine wav file:
<img width="960" alt="sinwav" src="https://user-images.githubusercontent.com/56707932/68921802-d6f35380-07b4-11ea-9dc0-e18529e96b0c.PNG">

aplay  /dev/zero:
<img width="960" alt="dev_zero" src="https://user-images.githubusercontent.com/56707932/68921807-d955ad80-07b4-11ea-9b1a-5f9d9bd586d2.PNG">

**Comment by xiulipan** [comment]
@Liviali155 Why put this issue in Linux? I through this would be FW issue.

@tlauda Please take a look here. 

Test with Logic Analysis and got some result: (above is capture with external USB card, below is raw data from LA result from SSP5 from UP2)
[12.txt](https://github.com/thesofproject/linux/files/3858475/12.txt)
![image](https://user-images.githubusercontent.com/37825714/69051190-9149bb80-0a3e-11ea-9cd0-a29283d6146f.png)

I think something wrong with the volume and PM is make chaos here. If I start 2 playback with very short delay. It seems the second playback will not have such issue.
![image](https://user-images.githubusercontent.com/37825714/69051446-4da38180-0a3f-11ea-9bd9-5fa00462dea5.png)

@Liviali155 @ClarexZhou Please try :
1.  with runtime PM disabled.
2.  with volume change with nocodec.

**Comment by xiulipan** [comment]
Test with nocodec with `aplay -Dhw:0,5 ~/audio_files/sine_48k_16bit_stereo_30sec.wav -vvv &  arecord -Dhw:0,5 -f dat test1118_3.wav -vvv`

Now can get similar issue with nocodec also, the key is the playback need to be fresh start. Volume is not set right and need some ramp to set to the right volume. @tlauda @singalsu Any idea about how to avoid such issue at a fresh start playback?

![image](https://user-images.githubusercontent.com/37825714/69051823-62344980-0a40-11ea-93d0-bd54afe72122.png)

**Comment by singalsu** [comment]
Volume in firmware is initialized to 0 dB (unity gain) if it is within the defined min/max range. If the playback starts with very loud output that gets attenuated the driver has not sent to firmware a volume min command before playback start. Now apparently a volume command for the actual volume is sent sometime in the beginning of playback. The driver ramps the volume from previous request (maybe 0 dB) to current  request. It does not know that it should ramp up from min volume in this case if the driver has not issued the command.

(Did I understood your observations correctly?)

**Comment by xiulipan** [comment]
@singalsu I think so.

Here is my question, if we find we are restore a volume change, do we still need to make the ramp up/down? For this case, we should make the volume start as the restore volume right?

**Comment by Liviali155** [comment]
> @Liviali155 Why put this issue in Linux? I through this would be FW issue.
> 
> @tlauda Please take a look here.
> 
> Test with Logic Analysis and got some result: (above is capture with external USB card, below is raw data from LA result from SSP5 from UP2)
> [12.txt](https://github.com/thesofproject/linux/files/3858475/12.txt)
> ![image](https://user-images.githubusercontent.com/37825714/69051190-9149bb80-0a3e-11ea-9cd0-a29283d6146f.png)
> 
> I think something wrong with the volume and PM is make chaos here. If I start 2 playback with very short delay. It seems the second playback will not have such issue.
> ![image](https://user-images.githubusercontent.com/37825714/69051446-4da38180-0a3f-11ea-9bd9-5fa00462dea5.png)
> 
> @Liviali155 @ClarexZhou Please try :
> 
> 1. with runtime PM disabled.
> 2. with volume change with nocodec.

1.with runtime PM disabled 
after disabled PM , error not seen
2. with volume change with nocodec.
tested with nocodec  with ```aplay -Dhw:0,5 -r48000 -c2 -fs16_le sine_48k_stereo_440.wav & arecord -Dhw:0,5 -c2 -r48000 -fs32 t1st.wav -vv -i```,error can reproduce,at the beginning the volume is very loud,but when change volume during aplay can not reproduce this issue

<img width="958" alt="Capture" src="https://user-images.githubusercontent.com/56707932/69127950-44212480-0ae6-11ea-91b2-e3cb7f99583a.PNG">

**Comment by singalsu** [comment]
> @singalsu I think so.
> 
> Here is my question, if we find we are restore a volume change, do we still need to make the ramp up/down? For this case, we should make the volume start as the restore volume right?

It would be better if the volume would ramp in playback start would be from mute/min to actual. Do you know why does suspending impact this (apparently this does not happen always)? I feel this is  a driver issue even if initializing to mute instead of 0 dB in volume new() might help in this case.

**Comment by dengyangchao** [comment]
GLK speaker pipeline(HW:0,0) can reproduce the issue. 
below two scenarios can reproduce:
1. PM runtime suspend ->playback via speaker
2. Playback via speaker -> suspend&resume

Platform: GLK Chrome with onboard codec DA7219 in I2S mode
Firmware: 65de27d2 https://github.com/thesofproject/sof/commits/master
Kernel: 7c7cc045 https://github.com/thesofproject/linux/commits/topic/sof-dev
Topology: file: tools/topology/sof-glk-da7219.tplg same as firmware


[dmesg.log](https://github.com/thesofproject/linux/files/3938501/dmesg.log)
[logger.log](https://github.com/thesofproject/linux/files/3938502/logger.log)

**Comment by singalsu** [comment]
@dengyangchao Could you test if SOF PR https://github.com/thesofproject/sof/pull/2109 helps with this issue. The PR is for capture but the part of PR that forces volume component to start with ramp from mute even if the driver does not request it should impact this issue as well.

**Comment by Liviali155** [comment]
> @dengyangchao Could you test if SOF PR [thesofproject/sof#2109](https://github.com/thesofproject/sof/pull/2109) helps with this issue. The PR is for capture but the part of PR that forces volume component to start with ramp from mute even if the driver does not request it should impact this issue as well.

Tried with thesofproject/sof#2109,error not seen on below platforms:
APL UP2 with codec PCM512X in I2S mode
GLK Chrome with onboard codec DA7219 in I2S mode

<img width="960" alt="Capture" src="https://user-images.githubusercontent.com/56707932/71141994-f0266e80-2250-11ea-8070-1287f75419d2.PNG">

<img width="951" alt="Capture1" src="https://user-images.githubusercontent.com/56707932/71141971-db49db00-2250-11ea-8f36-361dd17c6848.PNG">

**Comment by sinahuang** [comment]
Verified this issue with thesofproject/sof#2109 on below platform, issue can not be reproduced.
TGL-U RVP with codec ALC711 in SDW mode and  RT1308 in I2S mode 
CML-H Laptop with codec ALC711, 2 x RT1308, RT715 in SDW mode

**Comment by Liviali155** [comment]
closed it , PR [#2109](https://github.com/thesofproject/sof/pull/2109)  merged
